blog/_posts/2021-12-2-measuring-dev-env-health.md

```diff
@@ -0,0 +1,211 @@
+```
+
+(Tilt names each image as `resource-name:image-name`.)
+
+To see detailed info about the image build, use `-o=json`:
+
+```
+$ tilt get dockerimage tilt-site:tilt-site -o json
+{
+    ...
+    "spec": {
+        "context": "/home/nick/src/tilt.build",
+        "dockerfileContents": "FROM tilt-site-base\nADD ./src .\nADD healthcheck.sh .\nENTRYPOINT bundle exec jekyll serve --trace --config _config.yml,_config-dev.yml\n",
+        "ref": "tilt-site"
+    },
+    "status": {
+        "completed": {
+            "finishedAt": "2021-12-02T15:55:09.616393Z",
+            "startedAt": "2021-12-02T15:55:07.823548Z"
+        },
+        "ref": "localhost:38335/tilt-site:tilt-dc466fb6b52fa055",
+    ...
+}
+```
+
+The image object publishes when it started and when it finished the build.  You
+can even get fine-grained performance data about each stage of the image build,
+if I want to know which stage is the slowest.  But for this simple example,
+we're only going to collect the total build duration.
+
+Here's a little python code to collect the data:
+
+```python
+# Report duration times for each completed Docker Image build.
+
+# Fetch JSON representation of Docker Image status
+docker_image_list_json = subprocess.check_output(['tilt', 'get', 'dockerimage', '-o=json'])
+docker_image_list = json.loads(docker_image_list_json)
+builds = sorted(docker_image_list.get('items', []), key=lambda item: item['metadata']['name'])
+events = []
+for build in builds:
+  completed = build.get('status', {}).get('completed', None)
+  if not completed:
+    continue
+
+  # Parse timestamps.
+  start_time = datetime.datetime.strptime(completed['startedAt'], '%Y-%m-%dT%H:%M:%S.%fZ').replace(tzinfo=datetime.timezone.utc)
+  end_time = datetime.datetime.strptime(completed['finishedAt'], '%Y-%m-%dT%H:%M:%S.%fZ').replace(tzinfo=datetime.timezone.utc)
+  
+  # Add events.
+  events.append({
+    'data': {
+      'image_name': build['spec']['ref'],
+      'duration_ms': int((end_time - start_time).total_seconds() * 1000),
+      'kind': 'dockerimage',
+    },
+    'time': completed['finishedAt'],
+  })
+```
+
+Hooray! We have event data telling us how long each image build took.
+
+Next we need to report it.
+
+## Reporting the Data
+
+In this example, we're going to use [Honeycomb](https://honeycomb.io). Most observability services will work. 
+
+I chose Honeycomb for this example because:
+
+1) The free plan is more than enough to handle the volume of data we want to send.
+
+2) They have an ad-hoc [Events API](https://docs.honeycomb.io/api/events/) that
+   makes it easy to get started with your own data collection.
+
+Here's what the reporting loop looks like:
+
+```python
+api_key = 'MY-HONEYCOMB-API-KEY'
+dataset = 'MY-HONEYCOMB-DATASET'
+while True:
+  headers = {'X-Honeycomb-Team': api_key}
+  
+  events = collect_events() # ADD YOUR EVENT COLLECTION CODE HERE
+  client = http_client.HTTPSConnection("api.honeycomb.io", timeout=2)
```

Comment:
I know this isn't a complete example, but you might still put `import http.client as http_client` at the top since people are likely to copy-pasta this

Reply:
good point